File: Android/WatAreMyClasses/app/src/main/java/net/magmastone/NetworkInteraction/WatService.java
package net.magmastone.NetworkInteraction;

import net.magmastone.NetworkInteraction.Models.WatBuilding;
import net.magmastone.NetworkInteraction.Models.WatClass;
import net.magmastone.NetworkInteraction.Models.WatNode;

import java.util.List;

import retrofit.Callback;
import retrofit.http.GET;
import retrofit.http.Path;
import retrofit.http.Query;

/**
 * Created by alex on 8/6/15.
 *
 * This is the basic API implementation definition as required by Retrofit.
 *
 * Parameters and usage is identical to what's specified in docs/API.md. Contains a cross-platform overview.
 *
 *
 *
 */

public interface WatService {

        @GET("/getnextclass")
        void getNextClass(@Query("userid") String userID, @Query("token") String userToken, Callback<WatClass> cb); // Gets a given user's next class. Need both userID and token as taken from the barcode.

        @GET("/getschedule")
        void getSchedule(@Query("userid") String userID, @Query("token") String userToken,Callback<List<WatClass>> cb); // Gets a given user's entire day's schedule. Need both userID and token as taken from the barcode.


        @GET("/getpath")
        void getPath(@Query("node1") String nodeID1, @Query("node2") String nodeID2,Callback<List<WatNode>> cb); // Gets a path between two nodes. Requires no auth.

        @GET("/getclosestnode")
        void getClosestNode(@Query("lat") String lat, @Query("lon") String lon,Callback<WatNode> cb); // Gets the closest known node to the user. Requires no auth.

        @GET("/buildinglist")
        void buildingList(Callback<List<WatBuilding>> cb); // Gets all buildings on campus. SHOULD BE CACHED! Requires no auth.


}

File: iOS/WatAreMyClasses/WatAreMyClasses/WatService.m
//
//  WatService.m
//  WatAreMyClasses
//
//  Created by Alex Roth on 2015-08-07.
//  Copyright © 2015 Alex Roth. All rights reserved.
//

#import "WatService.h"
#import "WatNode.h"
#import "WatClass.h"
#import "WatBuilding.h"
#import "NetworkManager.h"
@implementation WatService
 // This is our abstractor class.
@synthesize nm;
+ (id)sharedService {
    static WatService *sharedService = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedService = [[self alloc] init];
    });
    return sharedService;
}
-(id) init{
    self=[super init];
    if (self){
        nm=[[NetworkManager alloc] initWithBaseURL:[NSURL URLWithString:@"http://ssvps.magmastone.net"]]; // Init with API endpoint.
        
    }
    return self;
}

//Get closest WatNode to the user.
- (void) getClosestNodeWithLat:(double) lat lon:(double) lon completion:(void (^) (OVCResponse *resp, NSError *error))complete {
    [nm GET:@"/getclosestnode" parameters:@{@"lat":[[NSNumber numberWithDouble:lat] stringValue],@"lon":[[NSNumber numberWithDouble:lon] stringValue]} completion:complete];
}

//Get a path between two given NodeIDs. Returns an NSArray of WatNodes.
-(void) getPathFromNode:(NSString*)node1 toNode:(NSString*)node2 completion:(void (^) (OVCResponse *resp, NSError *error))complete {
    [nm GET:@"/getpath" parameters:@{@"node1":node1,@"node2":node2} completion:complete];
}

//Get schedule for a given UserID and token. Get those values from the QR code. Returns an NSArray of WatClasses.
-(void)getScheduleforUser:(NSString*)userID andToken:(NSString*)token completion:(void (^) (OVCResponse *resp, NSError *error))complete{
    [nm GET:@"/getschedule" parameters:@{@"userid":userID,@"token":token} completion:complete];
}

//Get user's next class. Pass userID and token from QR code. Returns a WatClass.
-(void)getNextClassForUser:(NSString*)userID andToken:(NSString*)token completion:(void (^) (OVCResponse *resp, NSError *error))complete{
    [nm GET:@"/getnextclass" parameters:@{@"userid":userID,@"token":token} completion:complete];
}

//Get list of all buildings. Should be cached client and server-side.
-(void)getBuildings:(void (^) (OVCResponse *resp, NSError *error))complete{
    [nm GET:@"/buildinglist" parameters:nil completion:complete];
}

@end
